rhombus-lib/rhombus/private/amalgam/error.rhm

```diff
@@ -0,0 +1,164 @@
+#lang rhombus/private/amalgam/core
+import:
+  "core-meta.rkt" open
+  "maybe.rhm" open
+  lib("racket/base.rkt")
+
+use_static
+
+export:
+  error
+
+namespace error:
+  export:
+    Who
+
+    // constructs full message without raising it
+    message
+
+    // builders of main message test
+    annot_msg
+
+    Clause
+
+    // clause builders
+    text
+    val
+    vals
+    annot
+
+    // helper
+    reindent
+
+annot.macro 'Who':
+  'ReadableString || Symbol || Identifier || Operator'
+
+class Clause(msg :: String):
+  nonfinal
+
+// primary `error` function
+// (not backward-comptible with current `error` due to `~who` keyword argument)
+fun error(~srcloc: srcloc :: maybe(Srcloc) = #false,
+          ~who: who :: maybe(error.Who) = #false,
+          ~realm: realm :: Symbol = #'rhombus,
+          // make things like `~exn: Exn.Fail.User` easy; for more complex
+          // exception constructions, use `error.message`, instead, to build
+          // just the message text
+          ~exn: exn :: (Function.of_arity(2)
+                          // also gets srcloc list
+                          || Function.of_arity(3)
+                          // also gets clause list
+                          || Function.of_arity(4)) = Exn.Fail,
+          // intended to be a single short line
+          msg :: ReadableString,
+          // if non-empty, ";" is added to the end of `message`, and
+          // each `detail` is shown on its own line with a 1-space prefix
+          ~details: [detail :: ReadableString, ...] = [],
+          // intended to be constructed by `error.val`, `error.annot`,
+          // or `error.text`; each `clause` starts its own line with
+          // a 2-space prefix:
+          clause :: error.Clause,
+          ...):
+  let msg = message(~srcloc: srcloc,
+                    ~who: who,
+                    ~realm: realm,
+                    msg,
+                    ~detail: [detail, ...],
+                    clause, ...)
+  let marks = Continuation.Marks.current()
+  let e:
+    match exn
+    | exn :: Function.of_arity(4): exn(msg, marks, [srcloc], [clause, ...])
+    | exn :: Function.of_arity(3): exn(msg, marks, [srcloc])
+    | exn :: Function.of_arity(2): exn(msg, marks)
+  throw e
+
+// exported as `error.message`, same arguments as `error` except for `~exn`
+fun message(~srcloc: srcloc :: maybe(Srcloc) = #false,
+            ~who: who :: maybe(error.Who) = #false,
+            ~realm: realm :: Symbol = #'rhombus,
+            message :: ReadableString,
+            ~detail: [detail :: ReadableString, ...] = [],
+            clause :: error.Clause,
+            ...) :~ String:
+   (if srcloc | Srcloc.to_report_string(srcloc) ++ ": " | "")
```

Comment:
This should be prepended to the string below, if I understand correctly.

Reply:
Right - thanks!